js/web/lib/wasm/jsep/webgpu/ops/gather-block-quantized.ts

```diff
@@ -0,0 +1,193 @@
+          }`;
+          } else {
+            return `indices_indices = ${output.indicesGet('output_indices', 'uniforms.gather_axis')};`;
+          }
+        })()};
+        var data_indices = ${data.type.indices}(0);
+        for (var i: u32 = 0; i < uniforms.gather_axis; i++) {
+          let index = ${output.indicesGet('output_indices', 'i')};
+          ${data.indicesSet('data_indices', 'i', 'index')};
+        }
+        var index_from_indices = ${indices.getByIndices('indices_indices')};
+        if (index_from_indices < 0) {
+          index_from_indices += ${inputShape[gatherAxis]};
+        }
+        ${data.indicesSet('data_indices', 'uniforms.gather_axis', 'u32(index_from_indices)')};
+        for (var i = uniforms.gather_axis + 1; i < ${outputShape.length}; i++) {
+          let index = ${output.indicesGet('output_indices', `i + ${indicesShape.length} - 1`)};
+          ${data.indicesSet('data_indices', 'i', 'index')};
+        }
+        let data_offset = ${data.indicesToOffset('data_indices')};
+        let data_index = data_offset % 8;
+        // Convert 4-bit packed data to 8-bit packed data.
+        let packed_4bit_quantized_data = ${data.getByOffset('data_offset / 8')};
+        let packed_8bit_quantized_data = (packed_4bit_quantized_data >> (4 * (data_index % 2))) & 0x0f0f0f0f;
+        let quantized_data_vec = ${isSigned ? 'unpack4xI8' : 'unpack4xU8'}(u32(packed_8bit_quantized_data));
+        let quantized_data = quantized_data_vec[data_index / 2];
+        var scale_indices = data_indices;
+        let quantize_axis_index = ${scales.indicesGet('data_indices', 'uniforms.quantize_axis')} / uniforms.block_size;
+        ${scales.indicesSet('scale_indices', 'uniforms.quantize_axis', 'quantize_axis_index')};
+        var scale = ${scales.getByIndices('scale_indices')};
+        ${(() => {
+          if (!zeroPoint) {
+            return 'var zero_point = 0';
+          } else {
+            return `
+              let zero_point_indices = scale_indices;
+              let zero_point_offset = ${zeroPoint.indicesToOffset('zero_point_indices')};
+              let zero_point_index = zero_point_offset % 8;
+              let packed_4bit_zero_points = ${zeroPoint.getByOffset('zero_point_offset / 8')};
+              let packed_8bit_zero_points = (packed_4bit_zero_points >> (4 * (zero_point_index % 2))) & 0x0f0f0f0f;
+              let zero_point_vec = ${isSigned ? 'unpack4xI8' : 'unpack4xU8'}(u32(packed_8bit_zero_points));
+              let zero_point = zero_point_vec[zero_point_index / 2];`;
+          }
+        })()};
+        let dequantized_data = ${tensorTypeToWsglValueType(outputType)}(quantized_data - zero_point) * scale;
+        ${output.setByOffset('global_idx', 'dequantized_data')};
+    }`;
+  };
+  return {
+    name: 'GatherBlockQuantized',
+    shaderCache: {
+      hint: `${attributes.cacheKey};${inputs.map((input, _) => input.dims.join('_')).join(';')}`,
```

Comment:
inputs[1] is input_ids - we should avoid those dims from being known in the shader.

Reply:
We are only adding the shape of the indices input to the hint. Not the indices themselves. If the shape of the indices changes then the output shape also would change. Please confirm.

Reply:
A part of the output_indices goes to indexing input indices and the rest goes to indexing input combined with index from input_indices.

Reply:
Removed input[1] from cacheKey